// angular-cropperjs/bundles/angular-cropperjs.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('cropperjs'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-cropperjs', ['exports', '@angular/core', 'cropperjs', '@angular/common'], factory) :
    (factory((global['angular-cropperjs'] = {}),global.ng.core,null,global.ng.common));
}(this, (function (exports,i0,Cropper,common) { 'use strict';

    Cropper = Cropper && Cropper.hasOwnProperty('default') ? Cropper['default'] : Cropper;

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AngularCropperjsService = (function () {
        function AngularCropperjsService() {
        }
        AngularCropperjsService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] },
        ];
        /** @nocollapse */
        AngularCropperjsService.ctorParameters = function () { return []; };
        /** @nocollapse */ AngularCropperjsService.ngInjectableDef = i0.defineInjectable({ factory: function AngularCropperjsService_Factory() { return new AngularCropperjsService(); }, token: AngularCropperjsService, providedIn: "root" });
        return AngularCropperjsService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var CropperComponent = (function () {
        function CropperComponent() {
            this.cropperOptions = {};
            this.export = new i0.EventEmitter();
            this.ready = new i0.EventEmitter();
            this.isLoading = true;
        }
        /**
         * @return {?}
         */
        CropperComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        /**
         * Image loaded
         * @param ev
         */
        /**
         * Image loaded
         * @param {?} ev
         * @return {?}
         */
        CropperComponent.prototype.imageLoaded = /**
         * Image loaded
         * @param {?} ev
         * @return {?}
         */
            function (ev) {
                var _this = this;
                //
                // Unset load error state
                this.loadError = false;
                /** @type {?} */
                var image = (ev.target);
                this.imageElement = image;
                //
                // Add crossOrigin?
                if (this.cropperOptions.checkCrossOrigin)
                    image.crossOrigin = 'anonymous';
                //
                // Image on ready event
                image.addEventListener('ready', function () {
                    //
                    // Emit ready
                    //
                    // Emit ready
                    _this.ready.emit(true);
                    //
                    // Unset loading state
                    //
                    // Unset loading state
                    _this.isLoading = false;
                    //
                    // Validate cropbox existance
                    if (_this.cropbox) {
                        //
                        // Set cropbox data
                        //
                        // Set cropbox data
                        _this.cropper.setCropBoxData(_this.cropbox);
                    }
                });
                /** @type {?} */
                var aspectRatio = NaN;
                if (this.settings) {
                    var _a = this.settings, width = _a.width, height = _a.height;
                    aspectRatio = width / height;
                }
                //
                // Set crop options
                // extend default with custom config
                this.cropperOptions = Object.assign({
                    aspectRatio: aspectRatio,
                    movable: false,
                    scalable: false,
                    zoomable: false,
                    viewMode: 1,
                    checkCrossOrigin: true
                }, this.cropperOptions);
                //
                // Set cropperjs
                if (this.cropper) {
                    this.cropper.destroy();
                    this.cropper = undefined;
                }
                this.cropper = new Cropper(image, this.cropperOptions);
            };
        /**
         * Image load error
         * @param event
         */
        /**
         * Image load error
         * @param {?} event
         * @return {?}
         */
        CropperComponent.prototype.imageLoadError = /**
         * Image load error
         * @param {?} event
         * @return {?}
         */
            function (event) {
                //
                // Set load error state
                this.loadError = true;
                //
                // Unset loading state
                this.isLoading = false;
            };
        /**
         * Export canvas
         * @param base64
         */
        /**
         * Export canvas
         * @param {?=} base64
         * @return {?}
         */
        CropperComponent.prototype.exportCanvas = /**
         * Export canvas
         * @param {?=} base64
         * @return {?}
         */
            function (base64) {
                var _this = this;
                /** @type {?} */
                var imageData = this.cropper.getImageData();
                /** @type {?} */
                var cropData = this.cropper.getCropBoxData();
                /** @type {?} */
                var canvas = this.cropper.getCroppedCanvas();
                /** @type {?} */
                var data = { imageData: imageData, cropData: cropData };
                /** @type {?} */
                var promise = new Promise(function (resolve) {
                    //
                    // Validate base64
                    if (base64) {
                        //
                        // Resolve promise with dataUrl
                        return resolve({
                            dataUrl: canvas.toDataURL('image/png')
                        });
                    }
                    canvas.toBlob(function (blob) { return resolve({ blob: blob }); });
                });
                //
                // Emit export data when promise is ready
                promise.then(function (res) {
                    _this.export.emit(Object.assign(data, res));
                });
            };
        CropperComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'angular-cropper',
                        template: "<!-- CROPPER WRAPPER -->\n<div class=\"cropper-wrapper\">\n\n    <!-- LOADING -->\n    <div class=\"loading-block\" *ngIf=\"isLoading\">\n        <div class=\"spinner\"></div>\n    </div>\n\n    <!-- LOAD ERROR -->\n    <div class=\"alert alert-warning\" *ngIf=\"loadError\">{{ loadImageErrorText }}</div>\n\n    <!-- CROPPER -->\n    <div class=\"cropper\">\n        <img #image alt=\"image\" [src]=\"imageUrl\" (load)=\"imageLoaded($event)\" (error)=\"imageLoadError($event)\" />\n    </div>\n</div>\n",
                        styles: [":host{display:block}.cropper img{max-width:100%;max-height:100%;height:auto}.cropper-wrapper{position:relative;min-height:80px}.cropper-wrapper .loading-block{position:absolute;top:0;left:0;width:100%;height:100%}.cropper-wrapper .loading-block .spinner{width:31px;height:31px;margin:0 auto;border:2px solid rgba(97,100,193,.98);border-radius:50%;border-left-color:transparent;border-right-color:transparent;-webkit-animation:425ms linear infinite cssload-spin;position:absolute;top:calc(50% - 15px);left:calc(50% - 15px);animation:425ms linear infinite cssload-spin}@-webkit-keyframes cssload-spin{to{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}@keyframes cssload-spin{to{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}/*!\n * Cropper.js v1.4.1\n * https://fengyuanchen.github.io/cropperjs\n *\n * Copyright 2015-present Chen Fengyuan\n * Released under the MIT license\n *\n * Date: 2018-07-15T09:54:43.167Z\n */.cropper-container{direction:ltr;font-size:0;line-height:0;position:relative;touch-action:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.cropper-container img{display:block;height:100%;image-orientation:0deg;max-height:none!important;max-width:none!important;min-height:0!important;min-width:0!important;width:100%}.cropper-canvas,.cropper-crop-box,.cropper-drag-box,.cropper-modal,.cropper-wrap-box{bottom:0;left:0;position:absolute;right:0;top:0}.cropper-canvas,.cropper-wrap-box{overflow:hidden}.cropper-drag-box{background-color:#fff;opacity:0}.cropper-modal{background-color:#000;opacity:.5}.cropper-view-box{display:block;height:100%;outline:#39f solid 1px;overflow:hidden;width:100%}.cropper-dashed{border:0 dashed #eee;display:block;opacity:.5;position:absolute}.cropper-dashed.dashed-h{border-bottom-width:1px;border-top-width:1px;height:calc(100% / 3);left:0;top:calc(100% / 3);width:100%}.cropper-dashed.dashed-v{border-left-width:1px;border-right-width:1px;height:100%;left:calc(100% / 3);top:0;width:calc(100% / 3)}.cropper-center{display:block;height:0;left:50%;opacity:.75;position:absolute;top:50%;width:0}.cropper-center:after,.cropper-center:before{background-color:#eee;content:' ';display:block;position:absolute}.cropper-center:before{height:1px;left:-3px;top:0;width:7px}.cropper-center:after{height:7px;left:0;top:-3px;width:1px}.cropper-face,.cropper-line,.cropper-point{display:block;height:100%;opacity:.1;position:absolute;width:100%}.cropper-face{background-color:#fff;left:0;top:0}.cropper-line{background-color:#39f}.cropper-line.line-e{cursor:ew-resize;right:-3px;top:0;width:5px}.cropper-line.line-n{cursor:ns-resize;height:5px;left:0;top:-3px}.cropper-line.line-w{cursor:ew-resize;left:-3px;top:0;width:5px}.cropper-line.line-s{bottom:-3px;cursor:ns-resize;height:5px;left:0}.cropper-point{background-color:#39f;height:5px;opacity:.75;width:5px}.cropper-point.point-e{cursor:ew-resize;margin-top:-3px;right:-3px;top:50%}.cropper-point.point-n{cursor:ns-resize;left:50%;margin-left:-3px;top:-3px}.cropper-point.point-w{cursor:ew-resize;left:-3px;margin-top:-3px;top:50%}.cropper-point.point-s{bottom:-3px;cursor:s-resize;left:50%;margin-left:-3px}.cropper-point.point-ne{cursor:nesw-resize;right:-3px;top:-3px}.cropper-point.point-nw{cursor:nwse-resize;left:-3px;top:-3px}.cropper-point.point-sw{bottom:-3px;cursor:nesw-resize;left:-3px}.cropper-point.point-se{bottom:-3px;cursor:nwse-resize;height:20px;opacity:1;right:-3px;width:20px}@media (min-width:768px){.cropper-point.point-se{height:15px;width:15px}}@media (min-width:992px){.cropper-point.point-se{height:10px;width:10px}}@media (min-width:1200px){.cropper-point.point-se{height:5px;opacity:.75;width:5px}}.cropper-point.point-se:before{background-color:#39f;bottom:-50%;content:' ';display:block;height:200%;opacity:0;position:absolute;right:-50%;width:200%}.cropper-invisible{opacity:0}.cropper-bg{background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAA3NCSVQICAjb4U/gAAAABlBMVEXMzMz////TjRV2AAAACXBIWXMAAArrAAAK6wGCiw1aAAAAHHRFWHRTb2Z0d2FyZQBBZG9iZSBGaXJld29ya3MgQ1M26LyyjAAAABFJREFUCJlj+M/AgBVhF/0PAH6/D/HkDxOGAAAAAElFTkSuQmCC)}.cropper-hide{display:block;height:0;position:absolute;width:0}.cropper-hidden{display:none!important}.cropper-move{cursor:move}.cropper-crop{cursor:crosshair}.cropper-disabled .cropper-drag-box,.cropper-disabled .cropper-face,.cropper-disabled .cropper-line,.cropper-disabled .cropper-point{cursor:not-allowed}"],
                        encapsulation: i0.ViewEncapsulation.None
                    },] },
        ];
        /** @nocollapse */
        CropperComponent.ctorParameters = function () { return []; };
        CropperComponent.propDecorators = {
            image: [{ type: i0.ViewChild, args: ['image',] }],
            imageUrl: [{ type: i0.Input }],
            settings: [{ type: i0.Input }],
            cropbox: [{ type: i0.Input }],
            loadImageErrorText: [{ type: i0.Input }],
            cropperOptions: [{ type: i0.Input }],
            export: [{ type: i0.Output }],
            ready: [{ type: i0.Output }]
        };
        return CropperComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AngularCropperjsModule = (function () {
        function AngularCropperjsModule() {
        }
        AngularCropperjsModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [CropperComponent],
                        exports: [CropperComponent]
                    },] },
        ];
        return AngularCropperjsModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.AngularCropperjsService = AngularCropperjsService;
    exports.CropperComponent = CropperComponent;
    exports.AngularCropperjsModule = AngularCropperjsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
